zmq/log/handlers.py

```diff
@@ -56,22 +56,23 @@ class PUBHandler(logging.Handler):
     (DEBUG,INFO,etc.), followed by any additional subtopics specified in the
     message by: log.debug("subtopic.subsub::the real message")
     """
-    root_topic=""
+
     socket = None
     
-    formatters = {
-        logging.DEBUG: logging.Formatter(
-        "%(levelname)s %(filename)s:%(lineno)d - %(message)s\n"),
-        logging.INFO: logging.Formatter("%(message)s\n"),
-        logging.WARN: logging.Formatter(
-        "%(levelname)s %(filename)s:%(lineno)d - %(message)s\n"),
-        logging.ERROR: logging.Formatter(
-        "%(levelname)s %(filename)s:%(lineno)d - %(message)s - %(exc_info)s\n"),
-        logging.CRITICAL: logging.Formatter(
-        "%(levelname)s %(filename)s:%(lineno)d - %(message)s\n")}
     
-    def __init__(self, interface_or_socket, context=None):
+    def __init__(self, interface_or_socket, context=None, root_topic='root'):
```

Comment:
I'm not sure 'root' makes the most sense as a default root topic. No default makes more sense to me. Could it be that this should be the logger name by default? I'm not sure that's feasible, though, since the Handler doesn't have a handle on the Logger object.

Reply:
I really like the idea of the topic following the logger name, but as you said that may not be feasible, especially with how PUBHandler is built. It's designed to build up the message topic from the root topic, then the log level, then a subtopic passed in with a :: in the message. Seems tricky to modify that without changing existing behavior.

After we get past this, I was considering adding a new handler class that did use the logger name as the topic, because then the topics could follow the package hierarchy, and you could subscribe to log notifications based on where a module is in the layout of your codebase. How neat would that be!

I'm pretty sure the logger name would be available to do that - it is one of the attributes of LogRecord objects. 

But for this, as it is today, I don't care too much about what we call it, but I do think it shouldn't be blank. 

When I first started with this class, I inadvertently set the attribute `roottopic` instead of the correct `root_topic`. I got no errors (another reason I dislike direct attribute access in object APIs) but also no messages. I finally set my subscription to a blank value to start getting anything. That was confusing especially for a ZMQ newbie like myself - I didn't realize until very recently that you could have a "wildcard" subscription. 

My thinking is that if the value is populated with a default value, client code like the log watcher I built can "just work" out of the box, by subscribing to an expected value. But, we could get to the same outcome if we defaulted to a blank value, and then my log watcher could just subscribe to the empty string. I can happily make that change if you'd prefer - I just wasn't sure if "wildcard" subscriptions were to be encouraged in the ZMQ world.

Oh, and the reason I choose the string 'root' is because that is the name of the root logger in the core Python library. Could easily be something else.

Reply:
Since this is a Handler and we can't feasibly match the logger name, let's leave the existing behavior of no default root topic. With tree-based subscriptions in zmq PUB-SUB, the 'root' is really the empty string, not the word 'root'. Lacking a topic really is a valid and reasonable default behavior, and I'd personally like to leave that.

> I just wasn't sure if "wildcard" subscriptions were to be encouraged in the ZMQ world.

Yeah, it's quite common and in my experience what is done almost all of the time.

Reply:
Okay, done! Let me know if you see anything I missed.